### PR TITLE
Improved floating point comparison for better numerical stability

### PR DESCRIPTION
**Pull request by varunagrawal**
Using [this](https://floating-point-gui.de/errors/comparison/) as a reference, this PR adds a new function `fpEqual` to check if floating point numbers are equal in a numerically stable way.

This function is then used as in `Matrix` and `Vector` equality (with absolute tolerance) functions.

---
This change is [Reviewable](https://reviewable.io/reviews/borglab/gtsam/245)

### CODE REVIEW

File: gtsam/base/Vector.cpp [hunk 1]
```diff
@@ -33,6 +34,29 @@ using namespace std;
 
 namespace gtsam {
 
+/* ************************************************************************* */
+bool fp_isequal(double a, double b, double epsilon) {
+  double DOUBLE_MIN_NORMAL = std::numeric_limits<double>::min() + 1.0;
+
+  // handle NaNs
+  if(std::isnan(a) ^ std::isnan(b)) {
```
[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/base/Vector.cpp]: The caret is the bitwise XOR operator.

**Comment by ProfFan** [comment]
I see two test faults in CI, so I'll postpone the review.

File: gtsam/base/Vector.cpp [hunk 2]
```diff
@@ -33,6 +34,29 @@ using namespace std;
 
 namespace gtsam {
 
+/* ************************************************************************* */
+bool fp_isequal(double a, double b, double epsilon) {
```
[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/base/Vector.cpp]: Yup. Consistency is our friend.

File: gtsam/base/Vector.cpp [hunk 3]
```diff
@@ -33,6 +34,29 @@ using namespace std;
...
+  double DOUBLE_MIN_NORMAL = std::numeric_limits<double>::min() + 1.0;
```
[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/base/Vector.cpp]: I don't understand this. Can you please elaborate some more?

**Comment by varunagrawal** [comment]
Another excellent resource: https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/

File: gtsam/base/Vector.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dellaert) [reply gtsam/base/Vector.cpp]: Just put that at start of function body so we don't have `blah::` throughout, which clutters up the code and makes it less readable.
BTW, while Google naming convention is snake_case for free functions, we do use camelCase in GTSAM. `fpEqual`.

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/base/Vector.cpp]: Done and done!

File: gtsam/base/Vector.h [hunk 4]
```diff
@@ -82,10 +82,12 @@ static_assert(
  * Numerically stable function for comparing if floating point values are equal
  * within epsilon tolerance.
  * Used for vector and matrix comparison with C++11 compatible functions.
- * Reference : https://floating-point-gui.de/errors/comparison/
+ * References:
```
Comment (dellaert) [inline gtsam/base/Vector.h]: This is an implementation comment that should also go into the c++ file.

File: gtsam/base/Vector.h [hunk 5]
```diff
@@ -25,11 +25,11 @@
 #define MKL_BLAS MKL_DOMAIN_BLAS
 #endif
 
-#include <cmath>
-#include <limits>
 #include <gtsam/global_includes.h>
 #include <Eigen/Core>
+#include <cmath>
```
Comment (dellaert) [inline gtsam/base/Vector.h]: Header bloat! These headers should  go in c++, as mentioned in an earlier comment.

Comment (varunagrawal) [reply gtsam/base/Vector.h]: These are already present in the cpp file, so just removed them from the header file.

File: gtsam/base/Vector.h [hunk 6]
```diff
@@ -75,6 +76,17 @@ static_assert(
   "Error: GTSAM was built against a different version of Eigen");
 #endif
 
+/**
+ * Numerically stable function for comparing if floating point values are equal
+ * within epsilon tolerance.
+ * Used for vector and matrix comparison with C++11 compatible functions.
+ * References:
+ * 1. https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/
```
Comment (dellaert) [inline gtsam/base/Vector.h]: Possible  misunderstanding: I meant *move* the references to header, not copy them.

File: gtsam/base/Vector.h [hunk 7]
```diff
@@ -75,6 +76,17 @@ static_assert(
   "Error: GTSAM was built against a different version of Eigen");
 #endif
 
+/**
+ * Numerically stable function for comparing if floating point values are equal
+ * within epsilon tolerance.
```
Comment (ProfFan) [inline gtsam/base/Vector.h]: I think it is good to add comments on the behavior when we encounter NaNs and infinitys.

Comment (dellaert) [reply gtsam/base/Vector.h]: agreed. Just recap very succinctly.

Comment (varunagrawal) [reply gtsam/base/Vector.h]: Good point. I'll add it now.

**Comment by varunagrawal** [comment]
Added a note on behavior wrt NaNs and Infs. Since there were no code changes, I'll merge once CI passes.

**Pull request merged**